### PR TITLE
vmi-under-test: Log in to serial console

### PR DESCRIPTION
**Pull request by RamLavi**
Try to log in to the VMI serial console, in order to perform the necessary commands required to perform the checkup.

The console code is based Kiagnose/kubevirt-dpdk-checkup's console package [0]

[0] https://github.com/kiagnose/kubevirt-dpdk-checkup/tree/614adcb89fab7518d8c8a513c61dab99e1a8fd2b/pkg/internal/checkup/executor/console

### CODE REVIEW

**Comment by RamLavi** [comment]
getting error:
```
2023/12/03 09:15:08 kubevirt-realtime-checkup failed: admission webhook "virtualmachineinstances-create-validator.kubevirt.io" denied the request: spec.domain.memory.hugepages must be requested when NUMA topology strategy is set in spec.domain.cpu.numa.guestMappingPassthrough
```

**Comment by RamLavi** [comment]
Passed on CNV4.14 cluster:
```
make e2e-test
mkdir -p /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache
podman run --rm \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup:/go/src/github.com/kiagnose/kubevirt-realtime-checkup:Z \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache:/root/.cache/go-build:Z \
	-v /home/ralavi/.kube/sno03-cnvqe2-rdu2:/root/.kube:Z,ro \
	--workdir /go/src/github.com/kiagnose/kubevirt-realtime-checkup \
	-e KUBECONFIG=/root/.kube/kubeconfig \
	-e TEST_NAMESPACE=realtime-checkup-1 \
	-e TEST_CHECKUP_IMAGE=quay.io/ramlavi/kubevirt-realtime-checkup:devel \
	-e VM_UNDER_TEST_CONTAINER_DISK_IMAGE=quay.io/kiagnose/kubevirt-realtime-checkup-vm:main \
	docker.io/library/golang:1.19.4-bullseye \
	go test -v ./tests/...
=== RUN   TestTests
Running Suite: Tests Suite - /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests
========================================================================================
Random Seed: 1701680940

Will run 1 of 1 specs
•

Ran 1 of 1 Specs in 101.987 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestTests (101.99s)
PASS
ok  	github.com/kiagnose/kubevirt-realtime-checkup/tests	101.999s
```

checkup logs:
```
$ oc logs -f job/rt-checkup
2023/12/04 09:09:07 kubevirt-realtime-checkup starting...
2023/12/04 09:09:07 Using the following config:
2023/12/04 09:09:07 	"vmUnderTestTargetNodeName": ""
2023/12/04 09:09:07 	"vmUnderTestContainerDiskImage": "quay.io/kiagnose/kubevirt-realtime-checkup-vm:main"
2023/12/04 09:09:07 	"oslatDuration": "10m0s"
2023/12/04 09:09:07 	"oslatLatencyThresholdMicroSeconds": "45µs"
2023/12/04 09:09:07 Waiting for VMI "realtime-checkup-1/rt-vmi-9jm2s" to boot...
2023/12/04 09:10:08 VMI "realtime-checkup-1/rt-vmi-9jm2s" had successfully booted
2023/12/04 09:10:08 Login to VMI under test...
2023/12/04 09:10:31 Trying to delete VMI: "realtime-checkup-1/rt-vmi-9jm2s"
2023/12/04 09:10:31 Waiting for VMI "realtime-checkup-1/rt-vmi-9jm2s" to be deleted...
2023/12/04 09:10:36 VMI "realtime-checkup-1/rt-vmi-9jm2s" was deleted successfully
```

**Comment by orelmisan** [comment]
Could you please rebase?
I want to see if the go bump in #44 had caused issues.

**Comment by RamLavi** [comment]
[Change](https://github.com/kiagnose/kubevirt-realtime-checkup/compare/1c13882845aeecfb7e9f0480f771864bd457710c..c248b2cd308cb85bd8d90f693dd913fbdcb216a0): Rebase.

**Comment by RamLavi** [comment]
Passed on CNV4.15 cluster:
```
make e2e-test
mkdir -p /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache
podman run --rm \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup:/go/src/github.com/kiagnose/kubevirt-realtime-checkup:Z \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache:/root/.cache/go-build:Z \
	-v /home/ralavi/.kube/sno03-cnvqe2-rdu2:/root/.kube:Z,ro \
	--workdir /go/src/github.com/kiagnose/kubevirt-realtime-checkup \
	-e KUBECONFIG=/root/.kube/kubeconfig \
	-e TEST_NAMESPACE=realtime-checkup-1 \
	-e TEST_CHECKUP_IMAGE=quay.io/ramlavi/kubevirt-realtime-checkup:devel \
	-e VM_UNDER_TEST_CONTAINER_DISK_IMAGE=quay.io/kiagnose/kubevirt-realtime-checkup-vm:main \
	docker.io/library/golang:1.20.12 \
	go test -v ./tests/...
=== RUN   TestTests
Running Suite: Tests Suite - /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests
========================================================================================
Random Seed: 1701857697

Will run 1 of 1 specs
•

Ran 1 of 1 Specs in 117.075 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestTests (117.08s)
PASS
ok  	github.com/kiagnose/kubevirt-realtime-checkup/tests	117.088s
```
checkup logs:
```
$ oc logs -f job/rt-checkup
2023/12/06 10:15:05 kubevirt-realtime-checkup starting...
2023/12/06 10:15:05 Using the following config:
2023/12/06 10:15:05 	"vmUnderTestTargetNodeName": ""
2023/12/06 10:15:05 	"vmUnderTestContainerDiskImage": "quay.io/kiagnose/kubevirt-realtime-checkup-vm:main"
2023/12/06 10:15:05 	"oslatDuration": "10m0s"
2023/12/06 10:15:05 	"oslatLatencyThresholdMicroSeconds": "45µs"
2023/12/06 10:15:06 Waiting for VMI "realtime-checkup-1/rt-vmi-rzdv2" to boot...
2023/12/06 10:16:11 VMI "realtime-checkup-1/rt-vmi-rzdv2" had successfully booted
2023/12/06 10:16:11 Login to VMI under test...
2023/12/06 10:16:38 Trying to delete VMI: "realtime-checkup-1/rt-vmi-rzdv2"
2023/12/06 10:16:38 Waiting for VMI "realtime-checkup-1/rt-vmi-rzdv2" to be deleted...
2023/12/06 10:16:48 VMI "realtime-checkup-1/rt-vmi-rzdv2" was deleted successfully
```

File: pkg/internal/checkup/checkup.go [hunk 1]
```diff
@@ -49,6 +49,7 @@ type Checkup struct {
 	client    kubeVirtVMIClient
 	namespace string
 	vmi       *kvcorev1.VirtualMachineInstance
+	results   status.Results
```
Comment (orelmisan) [inline pkg/internal/checkup/checkup.go:57]: Please consider squashing / moving this commit closer to the commit introducing the executor.

File: pkg/internal/checkup/executor/executor.go [hunk 2]
```diff
@@ -34,21 +34,21 @@ type vmiSerialConsoleClient interface {
 }
 
 type Executor struct {
-	client      vmiSerialConsoleClient
-	namespace   string
-	vmiUsername string
-	vmiPassword string
+	vmiSerialClient vmiSerialConsoleClient
```
Comment (orelmisan) [inline pkg/internal/checkup/executor/executor.go:40]: This change could be a part of the previous commit.

File: pkg/internal/checkup/executor/executor.go [hunk 3]
```diff
@@ -34,21 +34,21 @@ type vmiSerialConsoleClient interface {
...
-		vmiPassword: config.VMIPassword,
+		vmiSerialClient: client,
+		namespace:       namespace,
+		vmiUsername:     config.VMIUsername,
+		vmiPassword:     config.VMIPassword,
 	}
 }
 
-func (e Executor) Execute(ctx context.Context, vmiName string) (status.Results, error) {
+func (e Executor) Execute(ctx context.Context, vmiUnderTestName string) (status.Results, error) {
```
Comment (orelmisan) [inline pkg/internal/checkup/executor/executor.go:55]: Same.

File: pkg/internal/checkup/checkup.go [hunk 4]
```diff
@@ -23,6 +23,7 @@ import (
 	"context"
 	"fmt"
 	"log"
+	"strings"
```
Comment (orelmisan) [inline pkg/internal/checkup/checkup.go:26]: Please consider moving this commit to the beginning, so the flow of serial console changes will not have a context switch.

File: pkg/internal/checkup/checkup.go (hunk 1 above)
Comment (RamLavi) [reply pkg/internal/checkup/checkup.go:57]: it makes more sense to move it to just before the executor is introduced.
DONE

File: pkg/internal/checkup/checkup.go (hunk 4 above)
Comment (RamLavi) [reply pkg/internal/checkup/checkup.go:26]: DONE

File: pkg/internal/checkup/executor/executor.go (hunk 2 above)
Comment (RamLavi) [reply pkg/internal/checkup/executor/executor.go:40]: DONE

File: pkg/internal/checkup/executor/executor.go (hunk 3 above)
Comment (RamLavi) [reply pkg/internal/checkup/executor/executor.go:55]: DONE

File: pkg/internal/checkup/checkup.go [hunk 5]
```diff
@@ -210,3 +206,14 @@ func randomizeName(prefix string) string {
 func ObjectFullName(namespace, name string) string {
 	return fmt.Sprintf("%s/%s", namespace, name)
 }
+
+func CloudInit(username, password string) string {
```
Comment (orelmisan) [inline pkg/internal/checkup/checkup.go:224]: In the DPDK checkup there is s [unit test](https://github.com/kiagnose/kubevirt-dpdk-checkup/blob/614adcb89fab7518d8c8a513c61dab99e1a8fd2b/pkg/internal/checkup/vmi_test.go#L92) for this function (this is the reason its exported).
Please consider adding it.

File: pkg/internal/config/config.go [hunk 6]
```diff
@@ -35,6 +35,9 @@ const (
 )
 
 const (
+	VMIUsername = "cloud-user"
+	VMIPassword = "0tli-pxem-xknu" // #nosec
```
Comment (orelmisan) [inline pkg/internal/config/config.go:39]: Does the first commit compiles?
AFAIU, the first commit (the one adding cloud init data) uses these consts that are introduces here.

**Review by orelmisan — CHANGES_REQUESTED** [review]

File: pkg/internal/checkup/checkup.go (hunk 5 above)
Comment (RamLavi) [reply pkg/internal/checkup/checkup.go:224]: I think it got lost on the conflict resolution and I missed it. will fix

File: pkg/internal/config/config.go (hunk 6 above)
Comment (RamLavi) [reply pkg/internal/config/config.go:39]: nope. I will move it to the first commit

Comment (RamLavi) [reply pkg/internal/config/config.go:39]: DONE

File: pkg/internal/checkup/checkup.go (hunk 5 above)
Comment (RamLavi) [reply pkg/internal/checkup/checkup.go:224]: DONE

**Comment by RamLavi** [comment]
Passed on CNV4.15 cluster:
```
make e2e-test
mkdir -p /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache
podman run --rm \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup:/go/src/github.com/kiagnose/kubevirt-realtime-checkup:Z \
	-v /home/ralavi/go/src/github.com/kiagnose/kubevirt-realtime-checkup/_go-cache:/root/.cache/go-build:Z \
	-v /home/ralavi/.kube/sno03-cnvqe2-rdu2:/root/.kube:Z,ro \
	--workdir /go/src/github.com/kiagnose/kubevirt-realtime-checkup \
	-e KUBECONFIG=/root/.kube/kubeconfig \
	-e TEST_NAMESPACE=realtime-checkup-1 \
	-e TEST_CHECKUP_IMAGE=quay.io/ramlavi/kubevirt-realtime-checkup:devel \
	-e VM_UNDER_TEST_CONTAINER_DISK_IMAGE=quay.io/kiagnose/kubevirt-realtime-checkup-vm:main \
	docker.io/library/golang:1.20.12 \
	go test -v ./tests/...
=== RUN   TestTests
Running Suite: Tests Suite - /go/src/github.com/kiagnose/kubevirt-realtime-checkup/tests
========================================================================================
Random Seed: 1702382246

Will run 1 of 1 specs
•

Ran 1 of 1 Specs in 107.015 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestTests (107.02s)
PASS
ok  	github.com/kiagnose/kubevirt-realtime-checkup/tests	107.027s
```
checkup logs:
```
$ oc logs -f job/rt-checkup
2023/12/12 11:57:30 kubevirt-realtime-checkup starting...
2023/12/12 11:57:30 Using the following config:
2023/12/12 11:57:30 	"vmUnderTestTargetNodeName": ""
2023/12/12 11:57:30 	"vmUnderTestContainerDiskImage": "quay.io/kiagnose/kubevirt-realtime-checkup-vm:main"
2023/12/12 11:57:30 	"oslatDuration": "1m0s"
2023/12/12 11:57:30 	"oslatLatencyThresholdMicroSeconds": "45µs"
2023/12/12 11:57:30 Waiting for VMI "realtime-checkup-1/rt-vmi-d5q6t" to boot...
2023/12/12 11:58:36 VMI "realtime-checkup-1/rt-vmi-d5q6t" had successfully booted
2023/12/12 11:58:36 Login to VMI under test...
2023/12/12 11:58:56 Trying to delete VMI: "realtime-checkup-1/rt-vmi-d5q6t"
2023/12/12 11:58:56 Waiting for VMI "realtime-checkup-1/rt-vmi-d5q6t" to be deleted...
2023/12/12 11:59:06 VMI "realtime-checkup-1/rt-vmi-d5q6t" was deleted successfully

```

**Review by orelmisan — APPROVED** [review]
Thank you

**Pull request merged**